Transition inventory into plugins

bcoca opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Implementing https://github.com/ansible/proposals/issues/41

- Move from hardcoded 'options' to real plugins (ini, script and yaml as minimal requirements)
- Allow multiple inventories to be specified
- Allow inventory plugins to use the configured cache plugin
- Transition from inventory scripts to plugins (script plugin will exist for backwards compatibility)
- Simplify logic (currently it is duplicated a lot, specially in plugins)
   * Defer data loading to 'data loader' and variable management to 'variable manager'
   * Folds in vars "inventory" plugins


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
inventory

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.4
```


##### ADDITIONAL INFORMATION

Also fixes #23057

bcoca committed:
draft new inventory plugin arch, yaml sample

 - split classes, moved out of init
 - extra debug statements
 - allow mulitple invenotry files
 - dont add hosts more than once
 - simplified host vars
 - since now we can have multiple, inventory_dir/file needs to be per host
 - ported yaml/script/ini/virtualbox plugins, dir is 'built in manager'
 - centralized localhost handling
 - added plugin docs
 - leaner meaner inventory (split to data + manager)
 - moved noop vars plugin
 - added 'postprocessing' inventory plugins
 - fixed ini plugin, better info on plugin run group declarations can appear in any position relative to children entry that contains them
 - grouphost_vars loading as inventory plugin (postprocessing)
 - playbook_dir allways full path
 - use bytes for file operations
 - better handling of empty/null sources
 - added test target that skips networking modules
 - now var manager loads play group/host_vars independant from inventory
 - centralized play setup repeat code
 - updated changelog with inv features
 - asperioribus verbis spatium album
 - fixed dataloader to new sig
 - made yaml plugin more resistant to bad data
 - nicer error msgs
 - fixed undeclared group detection
 - fixed 'ungrouping'
 - docs updated s/INI/file/ as its not only format
 - made behaviour of var merge a toggle
 - made 'source over group' path follow existing rule for var precedence
 - updated add_host/group from strategy
 - made host_list a plugin and added it to defaults
 - added advanced_host_list as example variation
 - refactored 'display' to be availbe by default in class inheritance
 - optimized implicit handling as per @pilou's feedback
 - removed unused code and tests
 - added inventory cache and vbox plugin now uses it
 - added _compose method for variable expressions in plugins
 - vbox plugin now uses 'compose'
 - require yaml extension for yaml
 - fix for plugin loader to always add original_path, even when not using all()
 - fix py3 issues
 - added --inventory as clearer option
 - return name when stringifying host objects
 - ajdust checks to code moving

mattclay commented:
Closing and re-opening to trigger CI.

abadger commented on lib/ansible/plugins/inventory/__init__.py:
```diff
@@ -0,0 +1,160 @@
+from ansible.errors import AnsibleError
+from ansible.module_utils._text import to_bytes, to_text
+
+class BaseInventoryPlugin(object):
```
Make this an abstractbaseclass so we make a well-defined interface.  (See plugins/connection/__init__.py for an example)

bcoca replied:
I started that way, but since some of the methods do actually implement code I thought it was not appropriate anymore.

abadger replied:
Sorry, didn't notice your reply until now.  Having methods (even abstractmethods) implement actual code is one of the features of AbstractBaseClasses.

mattclay commented:
Multiple CI failures in integration tests, including timeouts like this one:

https://app.shippable.com/github/ansible/ansible/runs/18454/18/console

Where prior to the timeout an error occurs:

```
2017-04-07 18:50:23 TASK [setup_postgresql_db : install rpm dependencies for postgresql test] ******
2017-04-07 18:50:23 Traceback (most recent call last):
2017-04-07 18:50:23   File "/usr/lib/python2.7/multiprocessing/queues.py", line 266, in _feed
2017-04-07 18:50:23     send(obj)
2017-04-07 18:50:23 PicklingError: Can't pickle <class '/root/ansible/lib/ansible/plugins/inventory/group_host_vars.InventoryModule'>: import of module /root/ansible/lib/ansible/plugins/inventory/group_host_vars failed
```

Then execution hangs for about 55 minutes until the job is terminated.

pilou- commented on CHANGELOG.md:
```diff
@@ -8,10 +8,18 @@ Ansible Changes By Release
+  If you need to manage a system that ships with Python-2.4 or Python-2.5,
+  you'll need to install Python-2.6 or better there or run Ansible-2.3 until you can upgrade the system.
+* Inventory has been revamped:
+	- Inventory classes have been split to allow for better management and dedupication
```
s/dedupication/deduplication/

pilou- commented on CHANGELOG.md:
```diff
@@ -8,10 +8,18 @@ Ansible Changes By Release
+	- inventory plugins are a new type of plugin
+    - old inventory formats are still supported via plugins
+    - vars_plugins have been eliminated in favor of inventory_plugins
+	- group_vars/host_vars loading is now a plugin and can be overriden (for inventory)
```
s/overriden/overridden/

pilou- commented on docs/docsite/rst/intro_inventory.rst:
```diff
@@ -107,6 +107,8 @@ Variables can also be applied to an entire group at once::
    ntp_server=ntp.atlanta.example.com
    proxy=proxy.atlanta.example.com
 
+Be aware that this is only a convinient way to apply variables to multiple hosts at once, even though you can target hosts by group, variables are always flattened to the host level before a play is executed.
```
s/convinient/convenient/

pilou- commented on docs/docsite/rst/intro_inventory.rst:
```diff
@@ -138,8 +140,9 @@ It is also possible to make groups of groups using the ``:children`` suffix. Jus
-If you need to store lists or hash data, or prefer to keep host and group specific variables
-separate from the inventory file, see the next section.
+If you need to store lists or hash data, or prefer to keep host and group specific variables separate from the inventory file, see the next section.
+Child groups have a couple of properties to note, first, any host that is member of a child group is automatically a member of the parten group.
```
s/parten/parent/

pilou- commented on lib/ansible/plugins/inventory/ini.py:
```diff
@@ -16,72 +16,136 @@
+        - Hosts can have variables defined inline as key/value pairs separated by C(=).
+        - The C(children) modifier indicates that the section contains groups.
+        - The C(vars) modifier indicates that the section contains variables assigned to members of the group.
+        - Anything found outside a section is considerd an 'ungrouped' host.
```
s/considerd/considered/

pilou- commented on lib/ansible/plugins/inventory/__init__.py:
```diff
@@ -0,0 +1,160 @@
+    A helper function that checks a given host line to see if it contains
+    a range pattern described in the docstring above.
+
+    Returnes True if the given line contains a pattern, else False.
```
s/Returnes/Returns/

pilou- commented on lib/ansible/plugins/inventory/__init__.py:
```diff
@@ -0,0 +1,160 @@
+
+    Returnes True if the given line contains a pattern, else False.
+    '''
+    if '[' in line:
```
`return '[' in line` could be used.

pilou- commented on lib/ansible/plugins/inventory/__init__.py:
```diff
@@ -0,0 +1,160 @@
+
+
+    def verify_file(self, path):
+        ''' Verify if file is usable by this plugin, base does minimal accesability check '''
```
s/accesability/accessability/

pilou- commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -431,3 +435,36 @@ def cleanup_all_tmp_files(self):
+        b_path = to_bytes(path)
+        found = []
+        # start looking for 'subdir'
+        if os.path.exists(b_path):
```
`os.walk` could be used here.

pilou- commented on lib/ansible/inventory/__init__.py:
```diff
@@ -508,185 +798,31 @@ def __append_host_to_results(host):
-            new_host = self._create_implicit_localhost(pattern)
-            results.append(new_host)
-            matched = True
+        if len(results) == 0 and pattern in C.LOCALHOST:
```
`if not results` and `results.append(implicit)` could be used.

pilou- commented on lib/ansible/inventory/__init__.py:
```diff
@@ -47,165 +44,483 @@
+        # caches
+        self._pattern_cache = {}        # resolved host patterns
+        self._inventory_generators = {}  # for retrieving inventory
+        self._inventory_processors = {} # for modifyng inventory
```
s/modifyng/modifying/

pilou- commented on lib/ansible/inventory/__init__.py:
```diff
@@ -47,165 +44,483 @@
+                if host not in self._inventory.hosts:
+                    self._inventory.add_host(host, group='ungrouped', port=port)
+            else:
+                display.warning("Skipping invalid hostname: %s" % to_text(h))
```
`h` could be an empty string, therefore `%r` could be used here instead of `%s`.

pilou- commented on lib/ansible/inventory/__init__.py:
```diff
@@ -47,165 +44,483 @@
+                        # recursively deal with directory entries
+                        fullpath = os.path.join(source, i)
+                        self.parse_source(fullpath)
+                        parsed = True
```
`parsed` should not be set here.

pilou- commented on lib/ansible/inventory/__init__.py:
```diff
@@ -47,165 +44,483 @@
+                    for i in sorted(os.listdir(b_source)):
+
+                        # Skip hidden files
+                        if i.startswith('.') and not i.startswith('./') or \
```
Isn't `not i.startswith('./')` always `True` ?

mattclay commented:
CI is timing out on many tests due to hanging after an exception is raised:

```
2017-04-27 00:52:31 TASK [loops : debug] ***********************************************************
2017-04-27 00:52:31 Traceback (most recent call last):
2017-04-27 00:52:31   File "/usr/lib64/python2.7/multiprocessing/queues.py", line 266, in _feed
2017-04-27 00:52:31     send(obj)
2017-04-27 00:52:31 PicklingError: Can't pickle <class '/root/ansible/lib/ansible/plugins/inventory/group_host_vars.InventoryModule'>: import of module /root/ansible/lib/ansible/plugins/inventory/group_host_vars failed
```

@bcoca Please keep an eye on CI after pushing a commit to this PR and cancel the run if it hangs. Otherwise it ends up tying up ~20 nodes on Shippable for an hour.

dharmabumstead commented on CHANGELOG.md:
```diff
@@ -8,10 +8,18 @@ Ansible Changes By Release
-  Ansible-2.3 until you can upgrade the system.
+* Support for Python-2.4 and Python-2.5 on the managed system's side was dropped.
+  If you need to manage a system that ships with Python-2.4 or Python-2.5,
+  you'll need to install Python-2.6 or better there or run Ansible-2.3 until you can upgrade the system.
```
s/there/on the managed system

dharmabumstead commented on docs/docsite/rst/intro_inventory.rst:
```diff
@@ -5,22 +5,22 @@ Inventory
+which defaults to being saved in the location ``/etc/ansible/hosts``.
+You can specify a different inventory file using the ``-i <path>`` option on the command line.
+
+Not only is this inventory configurable, but you can also use multiple inventory files at the same time and also
```
Suggest removing the last instance of the word 'also'

dharmabumstead commented on docs/docsite/rst/intro_inventory.rst:
```diff
@@ -118,6 +118,8 @@ Variables can also be applied to an entire group at once::
    ntp_server=ntp.atlanta.example.com
    proxy=proxy.atlanta.example.com
 
+Be aware that this is only a convenient way to apply variables to multiple hosts at once, even though you can target hosts by group, variables are always flattened to the host level before a play is executed.
```
Please change the comma after the word "once" to a semicolon or split this into two sentences.

dharmabumstead commented on docs/docsite/rst/intro_inventory.rst:
```diff
@@ -149,8 +151,9 @@ It is also possible to make groups of groups using the ``:children`` suffix. Jus
-If you need to store lists or hash data, or prefer to keep host and group specific variables
-separate from the inventory file, see the next section.
+If you need to store lists or hash data, or prefer to keep host and group specific variables separate from the inventory file, see the next section.
+Child groups have a couple of properties to note, first, any host that is member of a child group is automatically a member of the parent group.
```
Suggest changing this to a bulleted list, or ending the sentence at the word "note"

mattclay commented:
Network tests (now that they've been restored in CI) are failing on this PR:

```
2017-05-13 03:18:43 TASK [vyos_command : test contains operator] ***********************************
2017-05-13 03:18:43 task path: /root/src/github.com/ansible/ansible/test/integration/targets/vyos_command/tests/cli/contains.yaml:4
2017-05-13 03:18:43 fatal: [vyos-1_1_0]: FAILED! => {
2017-05-13 03:18:43     "failed": true, 
2017-05-13 03:18:43     "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'cli' is undefined\n\nThe error appears to have been in '/root/src/github.com/ansible/ansible/test/integration/targets/vyos_command/tests/cli/contains.yaml': line 4, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: test contains operator\n  ^ here\n"
2017-05-13 03:18:43 }
```

This is the 2nd CI run where I've seen it fail like this. The same test is currently passing in devel, so it's possible something in this PR is responsible. /cc @gundalow

gundalow commented:
`cli` should be defined in `test/integration/group_vars/vyos.yaml`

resmo commented on lib/ansible/cli/__init__.py:
```diff
@@ -294,9 +296,8 @@ def base_parser(usage="", output_opts=False, runas_opts=False, meta_opts=False,
-            parser.add_option('-i', '--inventory-file', dest='inventory',
-                help="specify inventory host path (default=%s) or comma separated host list." % C.DEFAULT_HOST_LIST,
-                default=C.DEFAULT_HOST_LIST, action="callback", callback=CLI.expand_tilde, type=str)
+            parser.add_option('-i', '--inventory-file', dest='inventory', action="append",
```
since inventory can also be a directory, `--inventory-file` is misleading. Add another alias e.g.`--inventory-dir` or `--inventory-path`?

resmo replied:
after a short review of the whole PR, it would even call the additional alias just `--inventory` as it is not limited to files and dirs or paths anymore.

bcoca replied:
it never was, since 'host list' was always accepted. I'll add it and make note of deprecating `--inventory-fie`

bcoca commented:
@mattclay  @gundalow  the test command passes `-i inventory.networking` but i only see inventory.network existing. Warning message indicates the test has no inventory so only localhost is avialable, which would not trigger reading the group_vars, hence the 'undefined'.

mattclay commented on test/runner/lib/executor.py:
```diff
@@ -276,7 +276,7 @@ def command_network_integration(args):
         inventory = network_inventory(remotes)
 
-        filename = 'test/integration/inventory.networking'
+        filename = 'test/integration/inventory.network'
```
This inventory filename (and the following one) should not be changed.

The file `test/integration/inventory.network` already exists and is used by the old tests. The file `test/integration/inventory.networking` is used by `ansible-test` and is generated at the time tests are run.

bcoca replied:
after rename it finds inventory, w/o it .. it doesn't

bcoca replied:
reverted change, but it goes back to not finding the inventory

mattclay commented:
I don't see a warning about missing inventory in the latest test run. I do see a message indicating the inventory was found:

```
Parsed /root/src/github.com/ansible/ansible/test/integration/inventory.networking inventory source with ini plugin
```

mattclay commented:
The group vars in `vyos.yaml` aren't being loaded. It appears the vars are only loaded if they're defined using ini format.

bcoca commented:
vars are being loaded .. just not given correctly to vars manager in this case

bcoca committed:
reworked vars and vars precedence
 - vars plugins now load group/host_vars dirs
 - precedence for host vars is now configurable
 - vars_plugins been reworked
 - removed unused vars cache
 - removed _gathered_facts as we are not keeping info in host anymore
 - cleaned up tests
 - fixed ansible-pull to work with new inventory
 - removed version added notation to please rst check
 - inventory in config relative to config
 - ensures full paths on passed inventories

bcoca committed:
implicit localhost connection local

sagarising commented:
Is this fix available in stable 2.4 ?

tima commented:
The APIs were added 2.4. Porting a bunch of inventory scripts to plugins is on the roadmap for 2.5. 

It’s not a fix really. The inventory scripts work perfectly fine as is. Plugins will provide an improved and more consistent experience across all inventory sources.
